backfiller/backfiller/main.py

```diff
@@ -0,0 +1,227 @@
+
+	Based on based on restreamer.list_segments. We could just call restreamer.list_segments but this avoids HTTP/JSON overheads."""
+
+	path = os.path.join(base_dir, stream, variant, hour)
+	try:
+		return [name for name in os.listdir(path) if not name.startswith('.')]
+
+	except OSError as e:
+		if e.errno != errno.ENOENT:
+			raise
+		return []	
+
+
+def list_remote_hours(node, stream, variant, timeout=TIMEOUT):
+	"""Wrapper around a call to restreamer.list_hours."""
+	uri = '{}/files/{}/{}'.format(node, stream, variant)
+	resp = requests.get(uri, timeout=timeout)
+	return resp.json()
+
+
+def list_remote_segments(node, stream, variant, hour, timeout=TIMEOUT):
+	"""Wrapper around a call to restreamer.list_segments."""
+	uri = '{}/files/{}/{}/{}'.format(node, stream, variant, hour)
+	resp = requests.get(uri, timeout=timeout)
+	return resp.json()
+
+
+def get_remote_segment(base_dir, node, stream, variant, hour, missing_segment,
+			timeout=TIMEOUT):
+	"""Get a segment from a node.
+
+	Fetches stream/variant/hour/missing_segment from node and puts it in base_dir/stream/variant/hour/missing_segment. If the segment already exists locally, this does not attempt to fetch it."""
+
+	path = os.path.join(base_dir, stream, variant, hour, missing_segment)
+	# check to see if file already exists to avoid unnecessarily copying it
+	if os.path.exists(path):
+		return
+	
+	dir_name = os.path.dirname(path)
+	date, duration, _ = os.path.basename(path).split('-', 2)
+	temp_name = "-".join([date, duration, "temp", str(uuid.uuid4())])
+	temp_path = os.path.join(dir_name, "{}.ts".format(temp_name))
+	common.ensure_directory(temp_path)
+
+	try:
+		uri = '{}/segments/{}/{}/{}/{}'.format(node, stream, variant, hour, missing_segment)
+		resp = requests.get(uri, stream=True, timeout=timeout)
+
+		resp.raise_for_status()
+
+		with open(temp_path, 'w') as f:
+			for chunk in resp.iter_content(8192):
+				f.write(chunk)
+
+	#try to get rid of the temp file if an exception is raised.
+	except Exception:
+		if os.path.exists(temp_path):
+			os.remove(temp_path)
+		raise
+
+	common.rename(temp_path, path)
+
+
+def backfill(base_dir, stream, variants, hours=None, nodes=None, start=None,
+		 stop=None, order=None):
+	"""Loop over nodes backfilling from each.
+
+	Backfill from node/stream/variants to base_dir/stream/variants for each node in nodes. If nodes is None, use get_nodes() to get a list of nodes to backfill from. Passes hours, start, stop and order to backfill_node to control which hours are backfilled and in which order. By default all hours are backfilled. If backfilling from a node raises an exception, this just goes onto the next node."""
+
+	if nodes is None:
+		nodes = get_nodes()
+
+	#ideally do this in parallel
+	for node in nodes:
+		try:
+			backfill_node(base_dir, node, stream, variants, hours, start, stop, order=order)
+		except Exception as e:
+			print node, e
```

Comment:
use `logging.exception()` to log the error, eg. `logging.exception("Error while backfilling node {}".format(node))`. logging.exception will automatically include the traceback and error info.